### ISSUE TITLE
new-haddock doesn't understand --for-hackage

### ISSUE THREAD

**Issue by ezyang0** [haskell/cabal#4412]
**Summary.** `cabal haddock` supports a `--for-hackage` flag which is not currently understood by `cabal new-haddock`. Steps to reproduce:

```
$ cabal new-haddock --for-hackage
cabal: unrecognized 'new-haddock' option `--for-hackage'
```

Adding support for this is a matter of threading this value through the various configuration data structures inside Cabal. If you grep for `haddockForHackage` you can see the main call-sites where support needs to be added. You'll need to add a new field to `ElaboratedConfiguredPackage` to store the Haddock flag. Look at another Haddock flag to see all of the places where it needs to be wired up.

**How to test if you're successful.** Run `cabal new-haddock --for-hackage` on a simple, one package new-build project with `-v`, and verify that `Setup haddock` is being invoked with `--for-hackage`.

**What you will learn.** How command line flags and project configuration are propagated through new-build; gain familiarity with some of the moving parts of new-build implementation.

**Comment by bitrauser** [comment]
I'd like to work on this.
Here is my progress so far, what I added:

```
.//Distribution/Client/ProjectConfig/Legacy.hs:      haddockForHackage         = packageConfigHaddockForHackage, -- JOEL
.//Distribution/Client/ProjectConfig/Legacy.hs:      haddockForHackage    = packageConfigHaddockForHackage, -- JOEL
.//Distribution/Client/ProjectConfig/Types.hs:       packageConfigHaddockForHackage   :: Flag HaddockTarget, -- JOEL
.//Distribution/Client/ProjectConfig/Types.hs:instance Binary HaddockTarget -- JOEL
.//Distribution/Client/ProjectPlanning/Types.hs:import           Distribution.Simple.Setup (HaddockTarget) -- JOEL
.//Distribution/Client/ProjectPlanning/Types.hs:       elabHaddockForHackage     :: HaddockTarget, -- JOEL
.//Distribution/Client/ProjectPlanning/Types.hs:instance Binary HaddockTarget -- JOEL
.//Distribution/Client/ProjectPlanning.hs:        elabHaddockForHackage   = perPkgOptionFlag pkgid Cabal.ForDevelopment packageConfigHaddockForHackage --JOEL
```

Building using `cabal new-build cabal-install` succeeds, however running `dist-newstyle/build/cabal-install-2.1.0.0/build/cabal/cabal` with `new-haddock --for-hackage` still reports that `--for-hackage` is unrecognized.

I am currently trying to figure out what I missed.

**Comment by ezyang** [comment]
@bitrauser Great!

Probably the problem is I forgot to tell you that you also need to tell the command line flag parsing code about the new flag. The command line info struct is defined in `cabal-install/Distribution/Client/CmdHaddock.hs` but there is some indirection here: what you really need to look at is `installCommand` in `cabal-install/Distribution/Client/Setup.hs`; if you look at `haddockOptions` you can see `for-hackage` is not in the list. Once you add it the command should be available under the name `--haddock-for-hackage`.

**Comment by bitrauser** [comment]
Thank you! After adjusting this I was able to run `cabal new-haddock --haddock-for-hackage -v` successfully and it appeared that `Setup haddock` was called with `--for-hackage`, as intended.

The Travis check is failing however, reason being the UnitTest of `ProjectConfig.hs`:

```
tests/UnitTests/Distribution/Client/ProjectConfig.hs:441:13: error:

    • The constructor ‘PackageConfig’ should have 43 arguments, but has been given 42
```

Which makes sense since I added the field `packageConfigHaddockForHackage :: Flag HaddockTarget,`  to `PackageConfig`, if I understand correctly.

In that case the UnitTest `PackageConfig` requires an additional argument.

Issue #4044 could be relevant.
